soroban-env-host/src/host/crypto.rs

```diff
@@ -77,13 +84,129 @@ impl Host {
         })
     }
 
+    #[cfg(feature = "next")]
+    pub(crate) fn secp256r1_verify_signature(
+        &self,
+        verifying_key: &p256::ecdsa::VerifyingKey,
+        msg_hash: &Hash,
+        sig: &Signature<p256::NistP256>,
+    ) -> Result<(), HostError> {
+        let _span = tracy_span!("p256 verify");
+        self.charge_budget(ContractCostType::VerifyEcdsaSecp256r1Sig, None)?;
+        verifying_key
+            .verify_prehash(msg_hash.as_slice(), sig)
+            .map_err(|_| {
+                self.err(
+                    ScErrorType::Crypto,
+                    ScErrorCode::InvalidInput,
+                    "failed scep256r1 verification",
+                    &[],
+                )
+            })
+    }
+
+    #[cfg(feature = "next")]
+    pub(crate) fn secp256r1_decode_sec1_uncompressed_pubkey(
+        &self,
+        bytes: &[u8],
+    ) -> Result<p256::ecdsa::VerifyingKey, HostError> {
+        use sec1::point::Tag;
+        self.charge_budget(ContractCostType::Sec1DecodePointUncompressed, None)?;
+        // check and make sure the key was encoded in uncompressed format
+        let tag = bytes
+            .first()
+            .cloned()
+            .ok_or(sec1::Error::PointEncoding)
+            .and_then(Tag::from_u8)
+            .map_err(|_| {
+                self.err(
+                    ScErrorType::Crypto,
+                    ScErrorCode::InvalidInput,
+                    "invalid ECDSA public key",
+                    &[],
+                )
+            })?;
+        if tag != Tag::Uncompressed {
+            return Err(self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InvalidInput,
+                "invalid ECDSA public key",
+                &[],
+            ));
+        }
+
+        p256::ecdsa::VerifyingKey::from_sec1_bytes(bytes).map_err(|_| {
+            self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InvalidInput,
+                "invalid ECDSA public key",
+                &[],
+            )
+        })
+    }
+
+    #[cfg(feature = "next")]
+    pub(crate) fn secp256r1_public_key_from_bytesobj_input(
+        &self,
+        k: BytesObject,
+    ) -> Result<p256::ecdsa::VerifyingKey, HostError> {
+        self.visit_obj(k, |bytes: &ScBytes| {
+            self.secp256r1_decode_sec1_uncompressed_pubkey(bytes.as_slice())
+        })
+    }
+
+    // ECDSA functions
+    #[cfg(feature = "next")]
+    pub(crate) fn ecdsa_signature_from_bytes<C>(
+        &self,
+        bytes: &[u8],
+    ) -> Result<Signature<C>, HostError>
+    where
+        C: PrimeCurve + CurveArithmetic,
+        SignatureSize<C>: ArrayLength<u8>,
+    {
+        self.charge_budget(ContractCostType::DecodeEcdsaCurve256Sig, None)?;
```

Comment:
Are we sure that `Signature::try_from` performs a length check before doing anything?

The same goes for the other cases byte slice inputs.

Reply:
That's correct. Checking length compatibility is the first thing it does underneath. 
I have test cases for wrong lengths to all inputs.